SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp

```diff
@@ -1712,12 +1777,349 @@ FString ActorSystem::GetObjectNameFromRepState(const FSpatialObjectRepState& Rep
+		}
+
+		NetDriver->Connection->SendDeleteEntityRequest(EntityId, RETRY_UNTIL_COMPLETE);
+	}
+}
+
+void ActorSystem::SendComponentUpdates(UObject* Object, const FClassInfo& Info, USpatialActorChannel* Channel,
+									   const FRepChangeState* RepChanges, const FHandoverChangeState* HandoverChanges,
+									   uint32& OutBytesWritten)
+{
+	// SCOPE_CYCLE_COUNTER(STAT_SpatialSenderSendComponentUpdates);
+	Worker_EntityId EntityId = Channel->GetEntityId();
+
+	UE_LOG(LogActorSystem, Verbose, TEXT("Sending component update (object: %s, entity: %lld)"), *Object->GetName(), EntityId);
+
+	USpatialLatencyTracer* Tracer = USpatialLatencyTracer::GetTracer(Object);
+	ComponentFactory UpdateFactory(Channel->GetInterestDirty(), NetDriver, Tracer);
+
+	TArray<FWorkerComponentUpdate> ComponentUpdates =
+		UpdateFactory.CreateComponentUpdates(Object, Info, EntityId, RepChanges, HandoverChanges, OutBytesWritten);
+
+	TArray<FSpatialGDKSpanId> PropertySpans;
+	if (EventTracer != nullptr && RepChanges != nullptr
+		&& RepChanges->RepChanged.Num() > 0) // Only need to add these if they are actively being traced
+	{
+		FSpatialGDKSpanId CauseSpanId;
+		if (EventTracer != nullptr)
+		{
+			CauseSpanId = EventTracer->PopLatentPropertyUpdateSpanId(Object);
+		}
+
+		for (FChangeListPropertyIterator Itr(RepChanges); Itr; ++Itr)
+		{
+			GDK_PROPERTY(Property)* Property = *Itr;
+
+			EventTraceUniqueId LinearTraceId = EventTraceUniqueId::GenerateForProperty(EntityId, Property);
+			FSpatialGDKSpanId PropertySpan = EventTracer->TraceEvent(
+				FSpatialTraceEventBuilder::CreatePropertyChanged(Object, EntityId, Property->GetName(), LinearTraceId),
+				CauseSpanId.GetConstId(), 1);
+
+			PropertySpans.Push(PropertySpan);
+		}
+	}
+
+	// It's not clear if this is ever valid for authority to not be true anymore (since component sets), but still possible if we attempt
+	// to process updates whilst an entity creation is in progress, or after the entity has been deleted or removed from view. So in the
+	// meantime we've kept the checking and queuing of updates, along with an error message.
+	const bool bHasAuthority = SubView->HasAuthority(EntityId, SpatialConstants::SERVER_AUTH_COMPONENT_SET_ID);
+	if (!bHasAuthority)
+	{
+		UE_LOG(LogActorSystem, Warning,
+			   TEXT("Trying to send component update but don't have authority! Update will be queued and sent when authority gained. "
+					"entity: %lld"),
+			   EntityId);
+
+		// It may be the case that upon resolving a component, we do not have authority to send the update. In this case, we queue the
+		// update, to send upon receiving authority. Note: This will break in a multi-worker context, if we try to create an entity that
+		// we don't intend to have authority over. For this reason, this fix is only temporary.
+		TArray<FWorkerComponentUpdate>& UpdatesQueuedUntilAuthority = UpdatesQueuedUntilAuthorityMap.FindOrAdd(EntityId);
+		UpdatesQueuedUntilAuthority.Append(ComponentUpdates);
```

Comment:
Didn't you remove processing this above?

Reply:
yes but now I'm not sure that was the best plan, need to think about this- having trouble thinking through if this is valid or not

Reply:
with component sets we should only be trying to send updates if we have authority, there's no more waiting for authority due to dynamic components for example. 
But then the comment mentions sending updates to entities that don't exist- why would that happen? If the entity migrates away mid tick? Do we really want to queue up this update? When we check out the entity again the state of the entity may have changed because this can only happen if someone else gets authority, and our update may be invalid now, and we're not checking that here. It seems we should be trusting the logic which handles checking out new entities to deal with migrated entities that are coming back in the same way as entities that have migrated here for the first time.
And then in that case- what did this code ever do? What's the concrete bug that it fixes? Do you know, and will it still be a problem now?

Reply:
What about the time between a create entity request, and receiving the entity add? In that instance, we keep Unreal authority, but we don't have an entity we can make updates on, so they need to be queued.

Reply:
why do we keep unreal authority? Can we make the entity creation/actor spawning synchronous?

Reply:
To simplify usability. This worker *will* have authority, and especially in a layered scenario where you never expect to lose authority, it's easier for the user to not have to deal with this. For them there's no distinction in operating on the actor before you create the entity, and before you check the entity out.

Reply:
It's the reason we added OnReady as as solution that's easy to ignore. We can't make this synchronous so if you do ignore OnReady then things might go wrong.

So actors keep authority but you aren't supposed to do anything with that actor that might interact with another entity until the entity is created.

In the case where you are just using offloading and you don't care about worker recovery or snapshots then you can ignore OnReady safely (except in the case of sending rpcs to startup actors). So it specifically improves usability there, while increasingly flexibility but at the cost of a pitfall elsewhere.

Reply:
added back the pre-auth buffering

Reply:
Chatted about this offline - I was incorrect about us processing updates whilst the entity creation is in progress (we check against that inside IsReadyForReplication) - so seems safe to remove this queuing logic all together, and replace it with an error message.